### ISSUE TITLE
Crashed: MediaFactoryImpl::signaling EXC_BAD_ACCESS KERN_INVALID_ADDRESS 0x0000000000000028

### ISSUE THREAD

**Issue by aksthelion** [twilio/voice-quickstart-ios#288]
I'm getting following crash in production app. 
Voice sdk 5.0

I was reading similar issue reported earlier for Obj-C and older verion but can't figure out what was the solution. How to fix this issue?

I don't have any other log, this is all what I've from Crashlytics log and I can't reproduce it. It happened only once. 

Crashed: MediaFactoryImpl::signaling
EXC_BAD_ACCESS KERN_INVALID_ADDRESS 0x0000000000000028

Crashed: MediaFactoryImpl::signaling
0  TwilioVoice                    0x1021a701c _hidden#43195_ + 3154080
1  TwilioVoice                    0x1021a82bc _hidden#6299_ + 3158848
2  TwilioVoice                    0x1021a82bc _hidden#6299_ + 3158848
3  TwilioVoice                    0x10206947c _hidden#35461_ + 117 (__hidden#35500_:117)
4  TwilioVoice                    0x102069570 _hidden#35468_ + 114 (__hidden#35500_:114)
5  TwilioVoice                    0x101fee7c4 _hidden#29956_ + 1016 (__hidden#25450_:1016)
6  TwilioVoice                    0x10204a988 _hidden#31603_ + 185 (__hidden#25415_:185)
7  TwilioVoice                    0x1020fbb7c _hidden#39724_ + 2452480
8  TwilioVoice                    0x101d89174 _hidden#5063_ + 337584
9  TwilioVoice                    0x101d928d8 _hidden#5679_ + 376340
10 TwilioVoice                    0x101da8598 _hidden#5647_ + 465620
11 TwilioVoice                    0x101da8514 _hidden#7015_ + 465488
12 libsystem_pthread.dylib        0x19b7e9840 _pthread_start + 168
13 libsystem_pthread.dylib        0x19b7f19f4 thread_start + 8

**Comment by bchen-twilio** [comment]
Hi @aksthelion 

Thanks for contacting us. We will take a look.
It would be great if you can provide more information such as what actions were performed, in case this happens again.

-bobie

**Comment by aksthelion** [comment]
Hello, Unfortunately it's a random error and I've only one instance of this crash in life app. It can't be reproduced and I've no ways of knowing what actions were performed by the user when this crash happens.

**Comment by bchen-twilio** [comment]
Hi @aksthelion 

Sorry for my delay of responding back to you.
No worries, we will look into the crash and investigate the root cause.

-bobie

**Comment by bchen-twilio** [comment]
Hi @aksthelion 

Hope this finds you well. We are wondering if you are using the latest Voice iOS SDK (currently 5.2.1) and are you still receiving crash report like this from the field?

-bobie

**Comment by dekpient** [comment]
@bchen-twilio 

We just had a similar crash in production. Not sure if it's the exact same crash or not. Please let me know if I need to create a new issue.

Twilio SDK: 5.2.0
Model: iPhone 6
OS version: 12.4.5

```
Crashed: MediaFactoryImpl::signaling
EXC_BAD_ACCESS KERN_INVALID_ADDRESS 0x000031470499d24c
_hidden#39669_

Crashed: MediaFactoryImpl::signaling
0  TwilioVoice                    0x1015784e4 _hidden#39669_ + 2777156
1  TwilioVoice                    0x10155ca60 _hidden#40300_ + 2663872
2  libc++.1.dylib                 0x1b5fa5d60 std::__1::__call_once(unsigned long volatile&, void*, void (*)(void*)) + 160
3  TwilioVoice                    0x10153db58 rtc::FireAndForgetAsyncClosure<twilio::voice::CallImpl::disconnect()::$_23>::Execute() + 2537144
4  TwilioVoice                    0x1011b6690 _hidden#5042_ + 337584
5  TwilioVoice                    0x1011bfdf4 _hidden#5658_ + 376340
6  TwilioVoice                    0x1011d5abc _hidden#5626_ + 465628
7  TwilioVoice                    0x1011d5a38 _hidden#6994_ + 465496
8  libsystem_pthread.dylib        0x1b6a1d2c0 _pthread_body + 128
9  libsystem_pthread.dylib        0x1b6a1d220 _pthread_start + 44
10 libsystem_pthread.dylib        0x1b6a20cdc thread_start + 4
```

**Comment by bchen-twilio** [comment]
Hi @dekpient 

Sorry for our lack of responses here and thank you for reporting the crash.
We have triaged the issue and are actively fixing the crash. We will keep you posted when we have a new release.

Sorry for the inconvenience.

Best,
-bobie

**Comment by brownpixel** [comment]
@bchen-twilio 
any update on this? I've been seeing this crash recently

**Comment by bchen-twilio** [comment]
Hi @brownpixel 

Thanks for reaching out. Can you provide the stacktrace, steps to reproduce and SDK log (if any) when the crash happened in your application?

We are actively working on this issue and will be providing a fix in the recent SDK release.
Sorry for the inconveniences.

**Comment by brownpixel** [comment]
@bchen-twilio unfortunately no repro steps. However based on logging it looks like it tends to happen randomly shortly after (ranges from 1-5 seconds) after connecting to a room and receiving another participant's video track.

Hope that helps and looking forward to the fix

**Comment by bchen-twilio** [comment]
@brownpixel thanks for the update. Although the Voice and Video iOS SDKs share similar media core stack, the work flow of connecting to a room or call could be slightly different especially that the Video SDK is more configurable with tracks. It would be great to create an issue and provide some more detail about the video app in the [Twilio Video iOS repo](https://github.com/twilio/twilio-video-ios/issues
).

Thanks,
bobie

**Comment by dekpient** [comment]
@bchen-twilio We're seeing a consistent pattern of this crash in our production app. We couldn't really reproduce the error 3 in development – not sure why – maybe you can try.

Basically, our app receives and processes PushKit VoIP notification on the main thread and reports the call. 
When CallKit fails to report the call with "com.apple.CallKit.error.incomingcall error 3”, we release the `callInvite` object from our delegate by setting `self.callInvite = nil` and looks like the TwilioVoice would crash then – see `[TVOCallInvite dealloc]` on the main thread stack from the crash report below.

Main Thread: com.apple.main-thread

```
0  libsystem_kernel.dylib         0x19700accc __psynch_cvwait + 8
1  libsystem_pthread.dylib        0x196f2d21c _pthread_cond_wait + 676
2  TwilioVoice                    0x102816fa8 _hidden#1505_ + 290872
3  TwilioVoice                    0x10282246c _hidden#5038_ + 337148
4  TwilioVoice                    0x102bb89c8 _hidden#40199_ + 2600940
5  TwilioVoice                    0x102bb8b20 _hidden#40201_ + 2601284
6  TwilioVoice                    0x102b8fea8 _hidden#39667_ + 2434252
7  TwilioVoice                    0x102b90574 _hidden#39679_ + 2435992
8  TwilioVoice                    0x102baea50 _hidden#40047_ + 2560116
9  TwilioVoice                    0x102baec00 _hidden#40049_ + 2560548
10 libobjc.A.dylib                0x196f38130 object_cxxDestructFromClass(objc_object*, objc_class*) + 116
11 libobjc.A.dylib                0x196f4af80 objc_destructInstance + 92
12 libobjc.A.dylib                0x196f51e44 _objc_rootDealloc + 52
13 TwilioVoice                    0x1027ab378 -[TVOCallInvite dealloc] + 3352
14 libobjc.A.dylib                0x196f533f8 objc_release + 136
15 mobile                         0x10216e2b0 __67-[Telephony reportIncomingCall:remoteHandle:localizedCallerName:]_block_invoke + 1147 (OUR CODE)
16 CallKit                        0x1a5411eb8 __62-[CXProvider reportNewIncomingCallWithUUID:update:completion:]_block_invoke_2 + 32
17 libdispatch.dylib              0x196ec4b7c _dispatch_call_block_and_release + 32
18 libdispatch.dylib              0x196ec5fd8 _dispatch_client_callout + 20
19 libdispatch.dylib              0x196ed1cc8 _dispatch_main_queue_callback_4CF + 968
20 CoreFoundation                 0x19719bcc8 __CFRUNLOOP_IS_SERVICING_THE_MAIN_DISPATCH_QUEUE__ + 16
21 CoreFoundation                 0x197196a24 __CFRunLoopRun + 1980
22 CoreFoundation                 0x197195f40 CFRunLoopRunSpecific + 480
23 GraphicsServices               0x1a1426534 GSEventRunModal + 108
24 UIKitCore                      0x19b321580 UIApplicationMain + 1940
25 mobile                         0x10206042c main + 14 (main.m:14)
26 libdyld.dylib                  0x197014e18 start + 4
```

Crashed Thread: MediaFactoryImpl::signaling
EXC_BAD_ACCESS KERN_INVALID_ADDRESS 0x0000000000000008

```
0  TwilioVoice                    0x102bbc180 _hidden#39735_ + 2615204
1  TwilioVoice                    0x102b945e4 _hidden#39734_ + 2452488
2  TwilioVoice                    0x102bb9d44 rtc::FireAndForgetAsyncClosure<twilio::voice::ListenerImpl::onRinging(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >)::$_2>::Execute() + 2605928
3  TwilioVoice                    0x1028224c4 _hidden#5042_ + 337236
4  TwilioVoice                    0x10282bbf4 _hidden#5658_ + 375940
5  TwilioVoice                    0x102841874 _hidden#5626_ + 465156
6  TwilioVoice                    0x1028417c4 _hidden#6994_ + 464980
7  libsystem_pthread.dylib        0x196f29840 _pthread_start + 168
8  libsystem_pthread.dylib        0x196f319f4 thread_start + 8
```

**Comment by bchen-twilio** [comment]
Thanks for providing additional debugging information @dekpient. We are actively working on solving the crash but the steps would definitely help us pinpoint the issue more easily.

**Comment by piyushtank** [comment]
@dekpient Apologies for the late response. 

We have a similar ticket for a crash in our backlogs with the similar stack trace and we are planning to prioritize the ticket in next sprint. To make sure what you are experiencing is the same crash, is it possible for you to share the crashlogs. You can post the crash logs here, or via a support ticket or can email it to me directly ptank@twilio.com 

Let me know if you have any questions.

**Comment by bchen-twilio** [comment]
Hi @dekpient 

Sorry for our lack of updates but we are happy to inform you that we have just released a new version of the Voice iOS SDK that includes the fix to crash during the call tear-down: https://www.twilio.com/docs/voice/voip-sdk/ios/changelog#541

Best,
-bobie

**Comment by dekpient** [comment]
@bobiechen-twilio We upgraded the SDK last week. The situation has definitely been improved – thank you for that. Unfortunately, we just had another crash in production. Please find the stack trace below.

TwilioVoice version 5.4.0
iOS  13.5.1 (17F80)
iPhone 11

```

Crashed: MediaFactoryImpl::signaling
0  TwilioVoice                    0x1036ef77c twilio::Logger::getModuleLogLevel(twilio::LogModule) const + 2808840
1  TwilioVoice                    0x1036f5410 twilio::media::IceConnectionActivityMonitor::getPeerConnectionStats() + 2832540
2  TwilioVoice                    0x10332665c rtc::AsyncInvoker::OnMessage(rtc::Message*) + 337236
3  TwilioVoice                    0x10332fd8c rtc::MessageQueue::Dispatch(rtc::Message*) + 375940
4  TwilioVoice                    0x103345a0c rtc::Thread::ProcessMessages(int) + 465156
5  TwilioVoice                    0x10334595c rtc::Thread::PreRun(void*) + 464980
6  libsystem_pthread.dylib        0x1a88d58fc _pthread_start + 168
7  libsystem_pthread.dylib        0x1a88dd9d4 thread_start + 8
```

**Comment by SimonRobinson** [comment]
@bobiechen-twilio  Another example of the crash. Please find the stack trace below. 

TwilioVoice version 5.4.0
IOS: 13.5.1 (17F80)
 iPhone 7

```
Crashed: MediaFactoryImpl::signaling
0  TwilioVoice                    0x1013eacf4 twilio::voice::SignalingClient::getSocket() + 2658688
1  TwilioVoice                    0x1013bacd0 twilio::voice::CallImpl::socketCreated() + 2462044
2  TwilioVoice                    0x1013e4918 rtc::FireAndForgetAsyncClosure<twilio::voice::ListenerImpl::onRinging(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >)::$_2>::Execute() + 2633124
3  TwilioVoice                    0x10104665c rtc::AsyncInvoker::OnMessage(rtc::Message*) + 337236
4  TwilioVoice                    0x10104fd8c rtc::MessageQueue::Dispatch(rtc::Message*) + 375940
5  TwilioVoice                    0x101065a0c rtc::Thread::ProcessMessages(int) + 465156
6  TwilioVoice                    0x10106595c rtc::Thread::PreRun(void*) + 464980
7  libsystem_pthread.dylib        0x185be7d98 _pthread_start + 156
8  libsystem_pthread.dylib        0x185beb74c thread_start + 8
```

**Comment by bobiechen-twilio** [comment]
Hi @dekpient @SimonRobinson 

Wanted to check in and see if you are still running into crashes with 5.4.1?

**Comment by dekpient** [comment]
@bobiechen-twilio We've been using 5.4.1 in production for the past 2 weeks and we haven't seen this crash since. We'll keep monitoring this crash and let you know if the situation changes. Thanks!

**Comment by dukhovnyi** [comment]
Hi @bobiechen-twilio, do you have any updates or some estimation regarding this issue? Could we help you with debugging or fixing this issue somehow?

**Comment by talhazafar914** [comment]
@bobiechen-twilio I encountered the same issue as reported in my Firebase Crashlytics. Please review it and inform me of any updates regarding this matter. Here's the stack trace

Crashed: MediaFactoryImpl::signaling 0x0x283df4c80
0  TwilioVoice                    0xd8d04 rtc::Thread::ProcessMessages(int) + 136
1  TwilioVoice                    0xd8ce0 rtc::Thread::ProcessMessages(int) + 100
2  TwilioVoice                    0xd9f94 rtc::Thread::PreRun(void*) + 60
3  libsystem_pthread.dylib        0x24d4 _pthread_start + 136
4  libsystem_pthread.dylib        0x1a10 thread_start + 8

**Comment by dipankadas** [comment]
@talhazafar914 Thanks for reaching out. A few questions:
1. What is the SDK version you are using? If that is not the latest, I would suggest please upgrade to the latest and see if the issue is present. We have recently fixed a couple of signaling thread crashes in the SDK.
2. If the issue still persists, we need to have steps for reproduction. Is it any way possible for you to provided steps for repro?
3. The stack trace is not very helpful here, so if you can repro this issue on a test device, then please let us know. Also please try to collect the SDK logs using the custom logger class if you can repro it in your test setup.

Thanks,
Dipankar.
